datahost-ld-openapi/src/tpximpact/datahost/ldapi/util/data_validation.clj

```
@@ -0,0 +1,200 @@
+         ;; NOTE: titles passed as a JS Array -> #{["foo" "bar"]} ?
+         ;; we throw if the array has more than one element.
+         fix-title (fn [schema] (update schema titles-key (partial unpack-column-title json-cols)))
+         indexed (set/index (into #{} (map fix-title) json-cols) [titles-key])
+         _ (assert (= (count json-cols) (count indexed))
+                   "column names in schema should be distinct")
+         m (reduce-kv (fn reducer [r k v]
+                        ;; we know there's only item per index key,
+                        ;; so we can safely unpack the value
+                        (let [col-name (get k titles-key)
+                              schema (extract-column-datatype (first v))]
+                          (assoc r col-name
+                                 (mu/update-properties schema assoc :dataset.column/name col-name))))
+                      {}
+                      indexed)]
+     (into [:tuple] (map #(get m %) (or column-names (keys m)))))))
+
+(defn row-schema->column-names
+  "Returns a seq of column names supported by the row schema."
+  [row-schema]
+  (map (comp :dataset.column/name m/properties) 
+       (m/children row-schema)))
+
+(defn- validate-found-columns
+  "Verifies that we could extract the required columns from the dataset.
+  Throws when it's not possible."
+  [schema column-names ds-column-names]
+  (let [schema-col-names (set column-names)
+        ds-col-names (set ds-column-names)]
+    (when (not= schema-col-names
+                (set/intersection ds-col-names schema-col-names))
+      (log/info "column mismatch" {:schema-column-names schema-col-names
+                                   :dataset-column-names ds-col-names})
+      (throw (ex-info "The expected columns are not in the dataset"
+                      {:schema-column-names schema-col-names
+                       :dataset-column-names ds-col-names
+                       :schema schema})))))
+
+(defn validate-dataset
```

This seems to be the root of the exception-throwing-for-flow-control.. one example I came across before for a more functional style that worked well was structuring it as a kind of reducer, accumulating error results (or failing fast) and passing that to each subsequent validator, there's an example here: https://github.com/Swirrl/defra-aims/blob/master/camc-app/src/camc_app/pipelines/validations.clj#L193

Maybe it would be more idiomatic to do something like that? Open to other suggestions too.

---

I can't view the linked repo. But I updated the code to not throw and return a map where caller has to check for potential `:explanation` key.